spec/controllers/devise/two_factor_authentication_controller_spec.rb

```diff
@@ -149,30 +149,77 @@ def index
 
       context 'when the user enters an invalid TOTP' do
         before do
-          patch :update, code: 'abc'
+          patch :update, code: 'abc', delivery_method: :totp
         end
 
         it 'increments second_factor_attempts_count' do
           expect(subject.current_user.reload.second_factor_attempts_count).to eq 1
         end
 
         it 're-renders the TOTP entry screen' do
-          expect(response).to render_template(:show_totp)
+          expect(response).to render_template(:confirm_totp)
         end
 
         it 'displays flash error message' do
           expect(flash[:error]).to eq t('devise.two_factor_authentication.attempt_failed')
         end
       end
 
-      context 'user requests a direct OTP via SMS and enters the direct OTP' do
+      context 'user requests a direct OTP via SMS' do
         before do
-          get :new, sms: true
-          patch :update, code: subject.current_user.direct_otp
+          get :new, delivery_method: :sms
         end
 
-        it 'redirects to the profile' do
-          expect(response).to redirect_to profile_path
+        it 'redirects to the confirmation screen' do
+          expect(response).to redirect_to(otp_confirm_path(delivery_method: :sms))
+        end
+
+        it 'does not render TOTP prompt' do
+          expect(response).to_not render_template(:confirm_totp)
+        end
+
+        context 'when user enters correct OTP' do
+          it 'redirects to the profile' do
+            patch :update, code: subject.current_user.direct_otp
+            expect(response).to redirect_to profile_path
+          end
+        end
+
+        context 'when user enters incorrect OTP' do
+          it 'displays flash error message' do
+            patch :update, code: 'rrrr', delivery_method: :sms
+            expect(flash[:error]).to eq t('devise.two_factor_authentication.attempt_failed')
+          end
+
+          it 'redirects to the confirmation screen' do
+            patch :update, code: 'rrrr', delivery_method: :sms
+            expect(response).to redirect_to(otp_confirm_path(delivery_method: :sms))
+          end
```

Comment:
Please combine expectations for tests that use the DB so as not to slow down the test suite. In this case, the redirection expectation would go inside the same `it` block as the flash expectation.

Alternatively, write your tests so they don't use the DB.